doc/dev/rpc-versioning.md

```diff
@@ -0,0 +1,214 @@
+# Runtime RPC versioning implementation notes
+
+This document describes the versioning protocol used to ensure two-way
+compatibility between different versions of the API.
+
+The approach is loosely inspired by the `Both_converts` model used by
+[Versioned_rpc](https://ocaml.janestreet.com/ocaml-core/latest/doc/async_rpc_kernel/Async_rpc_kernel/index.html#module-Versioned_rpc)
+in `Async`, in which both parties maintain a "menu" of supported RPC
+versions, which is used to negotiate a common protocol for each
+method.
+
+This is a working document and will be updated as the design evolves.
+
+## Terms
+
+- A **procedure** is a common term encompassing *notifications*
+    (one-way messages) and *requests* (a communication to which
+    a response is expected).
+
+- A **model** is the logical payload type for one direction of
+    a procedure. Note that this is a per-actor entity; the ultimate
+    goal of runtime RPC versioning is to allow clients and servers to
+    disagree on a model type without preventing them from interacting.
+
+- A **wire type** is the type sent between actors for one direction of
+    a procedure. This is negotiated for each procedure at
+    session-start time, and must be agreed upon by both actors. In normal
+    usage, one actor will use the model types for a procedure as the
+    wire types as well, with the counterparty performing the
+    conversions as necessary.
+
+- A **generation** of a procedure is the set of wire types
+    corresponding to each direction of a procedure, along with upgrade
+    and downgrade functions transforming the wire types to the model
+    types and vice versa. Each generation is associated with
+    a *version number*, which should be unique within a procedure.
```

Comment:
Shouldn't a generation also contain the serialisation and de-serialisation functions for the wire types?

Reply:
Good catch, added.